### DISCUSSION TITLE
Observability on duration of each middlewares

### DISCUSSION THREAD

**Discussion by mabead** [dotnet/aspnetcore#23661] [General]
I have an ASP.NET core 2.2 pipeline with about a dozen middlewares registered in the pipeline. In my first outer middleware I have code that basically calls the inner middleware and logs the time that it takes. It basically does this:

```csharp
var watch = Stopwatch.StartNew();
await _next.Invoke(context);
watch.Stop();
_logger.LogInformation(... watch.ElapsedMilliseconds);
```

This gives me `completePipelineDuration`.

Then, in my controller, I log the elapsed time spent in the controller as well.  This gives me `controllerDuration`.

I have then observed occurences where `completePipelineDuration` is close to 3 seconds and `controllerDuration` is close to 1 second. It therefore means that I have about 2 seconds that is being spend in my dozen of APS.NET Core middlewares.

So my question is this: is there a way to have metrics on how long it took to execute each middleware? I am trying to find out which piece of code is responsible for the performance issue.

**Comment by Tratcher** [comment]
Try this:

https://github.com/dotnet/aspnetcore/blob/df5269fea23f310fd7730b9eb0da5cc910cc1e6c/src/Middleware/MiddlewareAnalysis/samples/MiddlewareAnalysisSample/Startup.cs#L16
